README
-----------------------------------------
PINE.GPG - Bringing together PINE and GPG
-----------------------------------------

PINE.GPG is a message filter for (Al)pine, giving it the ability to interface
with GnuPG.  It lets the user verify signed incoming messages, decrypt
encrypted incoming messages, and encrypt and/or sign outgoing messages.

It was coded under Linux so it should compile and run on any modern distro
with gcc and glibc installed.  Other platforms are currently untested, but
any feedback and/or patches are welcome.

There are quite a few "PGP" type filters available for (Al)pine.  Here's some
information for helping to decide if PINE.GPG is right for you:

Some advantages over other PGP filters:

  Accounts for text outside of PGP message blocks.
  Able to handle multiple PGP messages per email.
  Fast, optimized C code.
  Easy to set up and use with no external dependencies or config files to edit.

Some disadvantages over other PGP filters:

  Only supports GnuPG, no support for any other PGP software.

See pine.gpg(1) for program usage and other info.

PINE.GPG ("this program") is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License version 2 as
published by the Free Software Foundation.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE. See the LICENSE file distributed with this program
for more details.

You should have received a copy of the GNU General Public License version 2
along with this program; if not, please write to:

  Free Software Foundation
  51 Franklin Street, Fifth Floor
  Boston, MA 02110-1301
  USA

Original source for this program available at:

  <http://www.absolutedigital.net/projects/pine.gpg/>
  <http://github.com/peake/pine.gpg/>

-- 


ChangeLog
---------------------------
Version 1.3.0 - 13 Sep 2014
---------------------------
 * Substantial rewrite of the existing codebase.
 * Data processing is now fully binary safe.
 * Use anonymous pipes instead of temporary files for internal IPC.
 * Replace system(3) calls with fork(2)+exec(3).
 * [NEW] Display filter can now handle multiple PGP messages per email as well
   as better handle text outside of PGP message blocks.
 * Command-line option to print version information (formerly -v) is now -V.
 * [NEW] Command-line option -v (formerly print version) is now for specifying
   verbosity level of GPG output.  Can be used twice for more verbosity.
   See gpg(1), option --verbose for additional information.
 * [SECURITY] Disable core dumps to help prevent leakage of sensitive data.
 * [OBSOLETE] Command-line option -t is still accepted (this may change in
   the future) but now ignored as we no longer use temporary files for IPC.
 * [DEPRECATED] Command-line option -e is now -s.  Option -e is still accepted
   as an alias of -s, but this may change in the future.
 * [NEW] Command-line option -r added for specifying a filter result file.
   See UPGRADING file for a caveat regarding this feature.
 * [NEW] Command-line option -k added for specifying a default signing key.
 * [NEW] Command-line options -S (sign), -E (encrypt), and -B (both) added to
   use sending filter without prompting.
 * [NEW] Command-line option -g added for specifying an alternative path at
   runtime to the GPG binary.
 * [NEW] Abort option added to the sending filter prompt.
 * [NEW] Sending filter prompt now allows for the selection of function without
   the need to press the ENTER key after character input.
 * Documentation updated.

Version 1.2.0 - Never released
------------------------------
 * Unfinished and abandoned.

Version 1.1.4 - 15 Nov 2004
---------------------------
 * encrypt.c: fixed an off-by-one memory allocation bug for recipient list

Version 1.1.3 - 09 Aug 2004
---------------------------
 * fixed typo in INSTALL

Version 1.1.2 - 08 Aug 2004
---------------------------
 * more optimizations
 * fixed typo in pine.gpg.1

Version 1.1.1 - 08 Aug 2004
---------------------------
 * code clean-ups and optimizations

Version 1.1.0 - 07 Aug 2004
---------------------------
 * encrypt, decrypt: FILE *stdin, *stdout, *stderr are constants so we need 
   another set of matching variables when re-assigning them; thanks to
   Andrew D. Hwang for catching this
 * main, encrypt, decrypt: __FUNCTION__ is a GCC extension of the 
   C99 __func__ so we should check if it's defined and define it if not
 * switched to an autoconf generated configure script
 * fixed a typo in the README

Version 1.0.0 - 04 Aug 2004
---------------------------
 * first public release

Version 0.4.3 - 03 Aug 2004
---------------------------
 * final code clean-up
 * minor bug fixes

Version 0.4.2 - 03 Aug 2004
---------------------------
 * documentation complete

Version 0.4.1 - 03 Aug 2004
---------------------------
 * better error handling in encrypt
 * bug fix for recipient list in encrypt

Version 0.4.0 - 01 Aug 2004
---------------------------
 * program code complete

Version 0.3.0 - 31 Jul 2004
---------------------------
 * first usable version


src/pinegpg.c
/*
 * Copyright (C) 2004-2014  Calvin E. Peake, Jr. <cp@absolutedigital.net>
 *
 * This file is part of PINE.GPG.
 *
 * PINE.GPG is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * PINE.GPG is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * LICENSE file distributed with PINE.GPG for more details.
 *
 * pinegpg.c - PINE.GPG entry point.
 * created 27 Jul 2004
 * origin  21 Apr 2004
 */

#include <sys/resource.h>
#include <sys/time.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <errno.h>

#include "pinegpg.h"
#include "display.h"
#include "sending.h"
#include "utility.h"

#include "config.h"

static const char *program_version = "1.3.1-rc1";

static void pr_usage(const char *program_name)
{
	printf("Usage: %s -d [-v...] [-r <file>] -i <file>\n"
	       "       %s -s [-v...] [-r <file>] -i <file> <recipient> "
	       "[<recipient>...]\n",
	       program_name, program_name);
}

static void exit_usage(const char *program_name)
{
	pr_usage(program_name);
	exit(EXIT_FAILURE); /* Always fail unless filtering completes OK. */
}

static void pr_version(void)
{
	printf("PINE.GPG %s\n"
	       "Copyright (C) 2004-2014  Calvin E. Peake, Jr.\n"
	       "Licensed under the GPLv2.\n", program_version);
}

static void exit_version(void)
{
	pr_version();
	exit(EXIT_FAILURE); /* Always fail unless filtering completes OK. */
}

static void pr_help(void)
{
	printf("\n"
"Options:\n"
"\n"
"  -d         Display filter mode: decrypt and/or verify.\n"
"  -s         Sending filter mode: encrypt and/or sign.\n"
"  -S         Sending filter mode: sign without prompting.\n"
"  -E         Sending filter mode: encrypt without prompting.\n"
"  -B         Sending filter mode: sign and encrypt without prompting.\n"
"  -i <file>  Input/output file.\n"
"  -r <file>  Result file for filtering status/errors.\n"
"  -g <path>  Specify an alternate path to the GPG binary.\n"
"  -k <key>   Specify the default signing key to use.\n"
"  -v         Have GPG be verbose in it's output.\n"
"  -h         Print program help (this screen) and exit.\n"
"  -V         Print program version and exit.\n"
	       "\n");
}

static void exit_help(const char *program_name)
{
	pr_version();
	pr_help();
	pr_usage(program_name);
	exit(EXIT_FAILURE); /* Always fail unless filtering completes OK. */
}

int main(int argc, char *argv[])
{
	char opt;
	struct rlimit limit;
	pinegpg_config config;

	config.mode = no_mode;
	config.input_file = NULL;
	config.result_file = NULL;
	config.rcpts = NULL;
	config.nr_rcpts = 0;
	config.gpg = GPG_PATH;
	config.default_key = NULL;
	config.verbose = 0;

	while ((opt = getopt(argc, argv, "BdEeg:hi:k:r:Sst:Vv")) != -1) {
		switch (opt) {
		case 'B':	/* sending filter, auto sign and encrypt */
			config.mode = both_mode;
			break;
		case 'd':	/* display filter (formerly decrypt) */
			config.mode = display_mode;
			break;
		case 'E':	/* sending filter, auto encrypt only */
			config.mode = encrypt_mode;
			break;
		case 'g':	/* gpg(1) */
			config.gpg = optarg;
			break;
		case 'h':	/* help */
			exit_help(argv[0]);
			break;
		case 'i':	/* input/output file */
			config.input_file = optarg;
			break;
		case 'k':	/* default key */
			config.default_key = optarg;
			break;
		case 'r':	/* result file */
			config.result_file = optarg;
			break;
		case 'S':	/* sending filter, auto sign only */
			config.mode = sign_mode;
			break;
		case 's':	/* sending filter */
		case 'e':	/* encrypt, deprecated */
			config.mode = sending_mode;
			break;
		case 't':	/* temporary directory, obsolete */
			break;
		case 'V':	/* program version */
			exit_version();
			break;
		case 'v':	/* gpg(1) verbose output */
			config.verbose++;
			break;
		default:
			exit_usage(argv[0]);
		}
	}

	if (config.input_file == NULL)
		exit_usage(argv[0]);

	if (optind < argc) {
		config.rcpts = malloc(sizeof (char *) * (argc - optind + 1));
		if (config.rcpts == NULL)
			die_x(EXIT_FAILURE, errno, config.result_file,
			      "Failed to create array for recipient list");

		for (; optind < argc; optind++, config.nr_rcpts++)
			config.rcpts[config.nr_rcpts] = argv[optind];
	}

	limit.rlim_cur = 0;
	limit.rlim_max = 0;
	if (setrlimit(RLIMIT_CORE, &limit))
		die_x(EXIT_FAILURE, errno, config.result_file,
		      "Failed to disable core dumps");

	if (config.mode == display_mode)
		display(&config);
	else if (config.mode >= sending_mode && config.nr_rcpts > 0)
		sending(&config);
	else
		exit_usage(argv[0]);

	/* We should never reach this point, but we will default to fail
	 * anyway because we never want the possibility of an unfiltered
	 * message getting sent out.
	 */
	return 1;
}


src/pinegpg.h
/*
 * Copyright (C) 2004-2014  Calvin E. Peake, Jr. <cp@absolutedigital.net>
 *
 * This file is part of PINE.GPG.
 *
 * PINE.GPG is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * PINE.GPG is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * LICENSE file distributed with PINE.GPG for more details.
 *
 * pinegpg.h - PINE.GPG entry point.
 * created 27 Jul 2004
 */

#ifndef PINEGPG_H
#define PINEGPG_H 1

#define BUF_SIZE 4096

typedef enum _program_mode {
	no_mode,
	display_mode,
	sending_mode,
	encrypt_mode,
	sign_mode,
	both_mode
} program_mode;

typedef struct _pinegpg_config {
	program_mode mode;
	char *input_file;
	char *result_file;
	char **rcpts;
	int  nr_rcpts;
	char *gpg;
	char *default_key;
	int  verbose;
} pinegpg_config;

#endif /* PINEGPG_H */


src/sending.c
/*
 * Copyright (C) 2004-2014  Calvin E. Peake, Jr. <cp@absolutedigital.net>
 *
 * This file is part of PINE.GPG.
 *
 * PINE.GPG is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * PINE.GPG is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * LICENSE file distributed with PINE.GPG for more details.
 *
 * sending.c - Sending filter.
 * created 27 Jul 2004
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <termios.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <fcntl.h>
#include <errno.h>

#include "pinegpg.h"
#include "utility.h"

/**
 * Sending filter for encrypting and/or signing.
 *
 * @param  config  The program configuration.
 * @return         Nothing.
 */
void sending(const pinegpg_config *config)
{
	int f, i, s, pout[2];
	int arg_idx = 0, nr_args = 14 + config->nr_rcpts * 2;
	pid_t pid;
	ssize_t bytes, wrote = 0, total = 0, gpg_out_size = BUF_SIZE;
	const size_t buf_size = gpg_out_size;
	char buf[buf_size], resp;
	char **gpg_args, *gpg_out, *gpg, *p;
	struct termios termio, termio_orig;

	const char *result_ok    = "Sending filter completed successfully.",
		   *result_abort = "Sending filter aborted.";

	gpg_args = malloc(sizeof (char *) * nr_args);
	if (gpg_args == NULL)
		die_x(EXIT_FAILURE, errno, config->result_file,
		      "Failed to create array for GPG arguments list");

	p = strrchr(config->gpg, '/');
	if (p == NULL)
		p = config->gpg;
	else
		p++;

	gpg = strdup(p);
	if (gpg == NULL)
		die_x(EXIT_FAILURE, errno, config->result_file,
		      "Failed to allocate memory for GPG process name");

	gpg_args[arg_idx++] = gpg;
	gpg_args[arg_idx++] = "--armor";
	gpg_args[arg_idx++] = "--set-filename";
	gpg_args[arg_idx++] = "";
	gpg_args[arg_idx++] = "--output";
	gpg_args[arg_idx++] = "-";

	if (config->default_key != NULL) {
		gpg_args[arg_idx++] = "--default-key";
		gpg_args[arg_idx++] = config->default_key;
	}

	if (config->verbose > 0)
		gpg_args[arg_idx++] = "--verbose";

	if (config->verbose > 1)
		gpg_args[arg_idx++] = "--verbose";

	switch (config->mode) {
	case sending_mode:
		if (tcgetattr(0, &termio) == -1)
			die_x(EXIT_FAILURE, errno, config->result_file,
			      "Failed to get terminal attributes");

		memcpy(&termio_orig, &termio, sizeof (struct termios));

		termio.c_lflag    &= ~ICANON;
		termio.c_cc[VMIN]  = 1;
		termio.c_cc[VTIME] = 0;

		if (tcsetattr(0, TCSANOW, &termio) == -1)
			die_x(EXIT_FAILURE, errno, config->result_file,
			      "Failed to set terminal attributes");

		for (i = 0; i < 1;) {
			printf("\n"
			       "  [PINE.GPG] (S)ign, (E)ncrypt, (B)oth, or "
			       "(A)bort (s/e/b/a)? ");
			resp = getchar();
			switch (resp) {
				case 's': case 'e': case 'b': case 'a': i++;
			}
		}
		printf("\n");

		if (tcsetattr(0, TCSANOW, &termio_orig) == -1)
			die_x(EXIT_FAILURE, errno, config->result_file,
			      "Failed to restore terminal attributes");
		break;
	case sign_mode:    resp = 's'; break;
	case encrypt_mode: resp = 'e'; break;
	case both_mode:    resp = 'b'; break;
	default:	   resp = 'a'; break;
	}

	switch (resp) {
	case 'a':
		die_x(EXIT_FAILURE, 0, config->result_file, result_abort);
	case 's':
		gpg_args[arg_idx++] = "--clearsign";
		break;
	case 'b':
		gpg_args[arg_idx++] = "--sign";
	case 'e':
		gpg_args[arg_idx++] = "--encrypt";
		for (i = 0; i < config->nr_rcpts; i++) {
			gpg_args[arg_idx++] = "--recipient";
			gpg_args[arg_idx++] = config->rcpts[i];
		}
	}

	gpg_args[arg_idx++] = config->input_file;
	gpg_args[arg_idx++] = NULL;

	if (pipe(pout) == -1)
		die_x(EXIT_FAILURE, errno, config->result_file,
		      "Failed to create pipe for stdout");

	pid = fork();
	if (pid == -1)
		die_x(EXIT_FAILURE, errno, config->result_file,
		      "Failed to create fork for GPG");

	if (pid == 0) {
		close(pout[0]);

		if (dup2(pout[1], 1) == -1)
			die_x(EXIT_FAILURE, errno, config->result_file,
			      "Failed to reassign GPG stdout to pipe");

		execv(config->gpg, gpg_args);

		die_x(127, errno, config->result_file, "Failed to execv(%s)",
		      config->gpg);
	}

	close(pout[1]);

	gpg_out = malloc(sizeof (char) * gpg_out_size);
	if (gpg_out == NULL)
		die_x(EXIT_FAILURE, errno, config->result_file,
		      "Failed to allocate buffer for GPG output");

	p = gpg_out;
	while ((bytes = read(pout[0], &buf, buf_size)) != 0) {
		if (bytes == -1) {
			if (errno == EINTR)
				continue;
			else
				die_x(EXIT_FAILURE, errno, config->result_file,
				      "GPG stdout read error");
		}

		while ((bytes + total) >= gpg_out_size) {
			gpg_out_size += buf_size;
			gpg_out = realloc(gpg_out,
					  sizeof (char) * gpg_out_size);
			if (gpg_out == NULL)
				die_x(EXIT_FAILURE, errno, config->result_file,
				      "Failed to increase GPG output buffer "
				      "size");
			if (gpg_out_size > (bytes + total))
				p = gpg_out + total;
		}

		memcpy(p, &buf, bytes);
		total += bytes;
		p += bytes;
	}

	do {
		while (waitpid(pid, &s, 0) == -1) {
			if (errno == EINTR)
				continue;
			else
				die_x(EXIT_FAILURE, errno, config->result_file,
				      "Failed to reap GPG child process %d",
				      pid);
		}
	} while (!WIFEXITED(s) && !WIFSIGNALED(s));

	if (WIFSIGNALED(s) && WTERMSIG(s))
		die_x(EXIT_FAILURE, 0, config->result_file,
		      "GPG process terminated by signal %d", WTERMSIG(s));

	if (WEXITSTATUS(s) > 0)
		die_x(EXIT_FAILURE, 0, config->result_file,
		      "GPG process exited with status %d", WEXITSTATUS(s));

	f = open(config->input_file, O_WRONLY | O_TRUNC);
	if (f == -1)
		die_x(EXIT_FAILURE, errno, config->result_file,
		      "Failed to open input file for writing");

	while ((bytes = write(f, gpg_out + wrote, total - wrote)) != 0) {
		if (bytes == -1) {
			if (errno == EINTR)
				continue;
			else
				die_x(EXIT_FAILURE, errno, config->result_file,
				      "Input file write error");
		}
		wrote += bytes;
	}

	if (total != wrote)
		die_x(EXIT_FAILURE, 0, config->result_file,
		      "Bytes wrote (%d) does not match output buffer size (%d)",
		      wrote, total);

	close(f);

	die_x(EXIT_SUCCESS, 0, config->result_file, result_ok);
}


src/utility.c
/*
 * Copyright (C) 2004-2014  Calvin E. Peake, Jr. <cp@absolutedigital.net>
 *
 * This file is part of PINE.GPG.
 *
 * PINE.GPG is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * PINE.GPG is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * LICENSE file distributed with PINE.GPG for more details.
 *
 * utility.c - Utility functions.
 * created 27 Jul 2004
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <string.h>
#include <stdarg.h>
#include <stdlib.h>
#include <stdio.h>
#include <fcntl.h>

/**
 * Print an error message to the result file if available, or stderr if not,
 * then terminate the program with the supplied status.
 *
 * @param  status  The status to pass to exit(3).
 * @param  errnum  If a non-zero value, then add strerror(3) to output.
 * @param  result  The path to the result file.
 * @param  format  A printf(3)-style format string.
 * @param  ...     A variable number of arguments for the format string.
 * @return         Nothing.
 */
void die_x(int status, int errnum, const char *result, const char *format, ...)
{
	int f, use_result = 0;
	va_list va;

	if (result != NULL) {
		f = open(result, O_WRONLY | O_CREAT | O_APPEND,
			 S_IRUSR | S_IWUSR);
		if (f != -1) {
			if (dup2(f, 2) == -1) {
				close(f);
				unlink(result);
			} else
				use_result = 1;
		}
	}

	if (!use_result) {
		fflush(stdout);
		fprintf(stderr, "  ");
	}

	fprintf(stderr, "[PINE.GPG] ");

	va_start(va, format);
	vfprintf(stderr, format, va);
	va_end(va);

	if (errnum != 0)
		fprintf(stderr, ": %s", strerror(errnum));

	fprintf(stderr, "\n");

	if (!use_result && status && status != 127) {
		fprintf(stderr, "\n  [PINE.GPG] Press ENTER to exit.");
		getchar();
	}

	exit(status);
}


src/display.c
/*
 * Copyright (C) 2004-2014  Calvin E. Peake, Jr. <cp@absolutedigital.net>
 *
 * This file is part of PINE.GPG.
 *
 * PINE.GPG is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * PINE.GPG is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * LICENSE file distributed with PINE.GPG for more details.
 *
 * display.c - Display filter.
 * created 27 Jul 2004
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/mman.h>
#include <sys/wait.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <fcntl.h>
#include <errno.h>

#include "pinegpg.h"
#include "utility.h"

/**
 * Decrypt and/or verify a PGP message.
 *
 * @param  input        The data to decrypt/verify.
 * @param  input_len    The size of the data in bytes.
 * @param  f            The file descriptor of our input-turned-output file.
 * @param  result_file  The path to our result file or NULL if none.
 * @param  gpg          The path to the gpg(1) binary.
 * @param  gpg_args     A list of arguments to be passed to gpg(1).
 * @return              Nothing.
 */
static void decrypt_message(const char *input, const int input_len,
			    const int f, const char *result_file,
			    const char *gpg, char * const *gpg_args)
{
	int i, s;
	int pin[2], pout[2], perr[2];
	static const int buf_size = BUF_SIZE;
	pid_t pid[2];
	char buf[buf_size], errmsg[64];
	ssize_t bytes, bytes_read, total;

	static const char *trl = "--[PINE.GPG]--------------------------"
				 "-------------------------------[TOP]--\n",
			  *grl = "--[PINE.GPG]--------------------------"
				 "-------------------------------[GPG]--\n",
			  *erl = "--[PINE.GPG]--------------------------"
				 "-------------------------------[END]--\n";

	if (pipe(pin) == -1)
		die_x(EXIT_FAILURE, errno, result_file,
		      "Failed to create pipe for stdin");

	pid[0] = fork();
	if (pid[0] == -1)
		die_x(EXIT_FAILURE, errno, result_file,
		      "Failed to create fork for feeder sub-process");

	if (pid[0] == 0) {
		close(pin[0]);

		total = 0;
		while ((bytes = write(pin[1], input + total,
				      input_len - total)) != 0) {
			if (bytes == -1) {
				if (errno == EINTR)
					continue;
				else
					die_x(127, errno, result_file,
					      "Feeder sub-process write error");
			}
			total += bytes;
		}

		close(pin[1]);

		if (total != input_len)
			die_x(127, 0, result_file, "Feeder sub-process bytes "
			      "wrote (%d) does not match input message size "
			      "(%d)", total, input_len);

		exit(EXIT_SUCCESS);
	}

	close(pin[1]);

	if (pipe(pout) == -1)
		die_x(EXIT_FAILURE, errno, result_file,
		      "Failed to create pipe for stdout");

	if (pipe(perr) == -1)
		die_x(EXIT_FAILURE, errno, result_file,
		      "Failed to create pipe for stderr");

	pid[1] = fork();
	if (pid[1] == -1)
		die_x(EXIT_FAILURE, errno, result_file,
		      "Failed to create fork for GPG");

	if (pid[1] == 0) {
		close(pout[0]);
		close(perr[0]);

		if (dup2(pin[0], 0) == -1)  /* stdin read from feeder */
			die_x(EXIT_FAILURE, errno, result_file,
			      "Failed to reassign GPG stdin to pipe");

		if (dup2(pout[1], 1) == -1) /* stdout write to parent */
			die_x(EXIT_FAILURE, errno, result_file,
			      "Failed to reassign GPG stdout to pipe");

		if (dup2(perr[1], 2) == -1) /* stderr write to parent */
			die_x(EXIT_FAILURE, errno, result_file,
			      "Failed to reassign GPG stderr to pipe");

		execv(gpg, gpg_args);

		die_x(127, errno, result_file, "Failed to execv(%s)", gpg);
	}

	close(pout[1]);
	close(perr[1]);

	if (mlock(&buf, buf_size) == -1)
		die_x(EXIT_FAILURE, errno, result_file,
		      "Failed to lock read buffer memory");

	total = 0;
	while ((bytes = write(f, trl + total, strlen(trl) - total)) != 0) {
		if (bytes == -1) {
			if (errno == EINTR)
				continue;
			else
				die_x(EXIT_FAILURE, errno, result_file,
				      "Input file write error");
		}
		total += bytes;
	}

	while ((bytes_read = read(pout[0], &buf, buf_size)) != 0) {
		if (bytes_read == -1) {
			if (errno == EINTR)
				continue;
			else
				die_x(EXIT_FAILURE, errno, result_file,
				      "GPG stdout read error");
		}

		total = 0;
		while ((bytes = write(f, buf + total,
				      bytes_read - total)) != 0) {
			if (bytes == -1) {
				if (errno == EINTR)
					continue;
				else
					die_x(EXIT_FAILURE, errno, result_file,
					      "Input file write error");
			}
			total += bytes;
		}
	}

	total = 0;
	while ((bytes = write(f, grl + total, strlen(grl) - total)) != 0) {
		if (bytes == -1) {
			if (errno == EINTR)
				continue;
			else
				die_x(EXIT_FAILURE, errno, result_file,
				      "Input file write error");
		}
		total += bytes;
	}

	while ((bytes_read = read(perr[0], &buf, buf_size)) != 0) {
		if (bytes_read == -1) {
			if (errno == EINTR)
				continue;
			else
				die_x(EXIT_FAILURE, errno, result_file,
				      "GPG stderr read error");
		}

		total = 0;
		while ((bytes = write(f, buf + total,
				      bytes_read - total)) != 0) {
			if (bytes == -1) {
				if (errno == EINTR)
					continue;
				else
					die_x(EXIT_FAILURE, errno, result_file,
					      "Input file write error");
			}
			total += bytes;
		}
	}

	/* Write these errors to the output file so that the user can see them.
	 * If we terminate with EXIT_FAILURE, then the MUA will not show any
	 * filtered text and the user will not be able to see the problem.
	 */
	for (i = 0; i < 2; i++) {
		do {
			while (waitpid(pid[i], &s, 0) == -1) {
				if (errno == EINTR)
					continue;
				else {
					snprintf(errmsg, sizeof (errmsg),
						 "  [PINE.GPG] Failed to reap "
						 "%s child process %d\n",
						 (i ? "GPG" : "feeder"),
						 pid[i]);
					write(f, errmsg, strlen(errmsg));
				}
			}
		} while (!WIFEXITED(s) && !WIFSIGNALED(s));

		if (WIFSIGNALED(s) && WTERMSIG(s)) {
			snprintf(errmsg, sizeof (errmsg),
				 "  [PINE.GPG] %s terminated by signal %d\n",
				 (i ? "GPG process" : "Feeder sub-process"),
				 WTERMSIG(s));
			write(f, errmsg, strlen(errmsg));
		}

		/* GPG exits with a status of one (1) if signature
		 * verification fails.  A status of greater than one (1)
		 * indicates a "real" error.
		 */
		if ((i == 0 && WEXITSTATUS(s) > 0) ||
		    (i == 1 && WEXITSTATUS(s) > 1)) {
			snprintf(errmsg, sizeof (errmsg),
				 "  [PINE.GPG] %s exited with status %d\n",
				 (i ? "GPG process" : "Feeder sub-process"),
				 WEXITSTATUS(s));
			write(f, errmsg, strlen(errmsg));
		}
	}

	total = 0;
	while ((bytes = write(f, erl + total, strlen(erl) - total)) != 0) {
		if (bytes == -1) {
			if (errno == EINTR)
				continue;
			else
				die_x(EXIT_FAILURE, errno, result_file,
				      "Input file write error");
		}
		total += bytes;
	}
}

/**
 * Display filter for decrypting and/or verifying signatures.
 *
 * @param  config  The program configuration.
 * @return         Nothing.
 */
void display(const pinegpg_config *config)
{
	int f;
	int arg_idx = 0, nr_args = 6;
	int pgp_begin_len, pgp_end_len, pgp_len;
	int pgp_msg_begin_len, pgp_msg_end_len, pgp_msg_len;
	int pgp_signed_begin_len, pgp_signed_end_len, pgp_signed_len;
	const char *e, *p, *pb, *pe, *pl;
	const char *pgp_begin, *pgp_end;
	char **gpg_args, *gpg, *input;
	ssize_t bytes, total, input_size;
	struct stat sbuf;

	const char *result_ok    = "Display filter completed successfully.",
		   *result_empty = "Display filter skipped empty input.";

	const char *pgp_msg_begin    = "-----BEGIN PGP MESSAGE-----\n",
		   *pgp_msg_end      = "-----END PGP MESSAGE-----\n",
		   *pgp_signed_begin = "-----BEGIN PGP SIGNED MESSAGE-----\n",
		   *pgp_signed_end   = "-----END PGP SIGNATURE-----\n";

	pgp_msg_begin_len = strlen(pgp_msg_begin);
	pgp_msg_end_len   = strlen(pgp_msg_end);
	pgp_msg_len       = pgp_msg_begin_len + pgp_msg_end_len;

	pgp_signed_begin_len = strlen(pgp_signed_begin);
	pgp_signed_end_len   = strlen(pgp_signed_end);
	pgp_signed_len       = pgp_signed_begin_len + pgp_signed_end_len;

	gpg_args = malloc(sizeof (char *) * nr_args);
	if (gpg_args == NULL)
		die_x(EXIT_FAILURE, errno, config->result_file,
		      "Failed to create array for GPG arguments list");

	p = strrchr(config->gpg, '/');
	if (p == NULL)
		p = config->gpg;
	else
		p++;

	gpg = strdup(p);
	if (gpg == NULL)
		die_x(EXIT_FAILURE, errno, config->result_file,
		      "Failed to allocate memory for GPG process name");

	gpg_args[arg_idx++] = gpg;

	if (config->verbose > 0)
		gpg_args[arg_idx++] = "--verbose";

	if (config->verbose > 1)
		gpg_args[arg_idx++] = "--verbose";
	/*
	 * This should be --decrypt for both decryption and signature
	 * verification.  Using --verify does not print out the verified
	 * (and unescaped) data.
	 */
	gpg_args[arg_idx++] = "--decrypt";
	gpg_args[arg_idx++] = "-";
	gpg_args[arg_idx++] = NULL;

	if (stat(config->input_file, &sbuf) == -1)
		die_x(EXIT_FAILURE, errno, config->result_file,
		      "Failed to get size of input file");

	input_size = sbuf.st_size;
	if (input_size == 0)
		die_x(EXIT_SUCCESS, 0, config->result_file, result_empty);

	input = malloc(sizeof (char) * input_size);
	if (input == NULL)
		die_x(EXIT_FAILURE, errno, config->result_file,
		      "Failed to allocate buffer for input file");

	f = open(config->input_file, O_RDWR);
	if (f == -1)
		die_x(EXIT_FAILURE, errno, config->result_file,
		      "Failed to open input file for read/write");

	total = 0;
	while ((bytes = read(f, input + total, input_size - total)) != 0) {
		if (bytes == -1) {
			if (errno == EINTR)
				continue;
			else
				die_x(EXIT_FAILURE, errno, config->result_file,
				      "Input file read error");
		}
		total += bytes;
	}

	if (total != input_size)
		die_x(EXIT_FAILURE, 0, config->result_file,
		      "Bytes read (%d) does not match input file size (%d)",
		      total, input_size);

	if (lseek(f, 0, SEEK_SET) == -1)
		die_x(EXIT_FAILURE, errno, config->result_file,
		      "Failed to seek to beginning of input file");

	if (ftruncate(f, 0) == -1)
		die_x(EXIT_FAILURE, errno, config->result_file,
		      "Failed to truncate input file");

	p = pe = pl = input;
	e = input + input_size;

	for (; p < e; p++) {
		if (*p != '-')
			continue;
		if ((p + pgp_signed_len) <= e &&
		    (p == input || *(p - 1) == '\n') &&
		    memcmp(p, pgp_signed_begin, pgp_signed_begin_len) == 0) {
			pgp_begin     = pgp_signed_begin;
			pgp_begin_len = pgp_signed_begin_len;
			pgp_end	      = pgp_signed_end;
			pgp_end_len   = pgp_signed_end_len;
			pgp_len       = pgp_signed_len;
		} else if ((p + pgp_msg_len) <= e &&
			   (p == input || *(p - 1) == '\n') &&
			   memcmp(p, pgp_msg_begin, pgp_msg_begin_len) == 0) {
			pgp_begin     = pgp_msg_begin;
			pgp_begin_len = pgp_msg_begin_len;
			pgp_end       = pgp_msg_end;
			pgp_end_len   = pgp_msg_end_len;
			pgp_len       = pgp_msg_len;
		} else
			continue;

		pb = p;
		p += pgp_begin_len;
		for (; p <= e; p++) {
			if (*p != '-')
				continue;
			if ((p + pgp_end_len) <= e &&
			    *(p - 1) == '\n' &&
			    memcmp(p, pgp_end, pgp_end_len) == 0) {
				while (pl != pb &&
				       (bytes = write(f, pl, pb - pl)) != 0) {
					if (bytes == -1) {
						if (errno == EINTR)
							continue;
						else
							die_x(EXIT_FAILURE,
							      errno,
							      config->result_file,
							      "Input file "
							      "write error");
					}
					pl += bytes;
				}
				p += pgp_end_len;
				pe = pl = p;
				decrypt_message(pb, p - pb, f,
						config->result_file,
						config->gpg, gpg_args);
				break;
			}
		}
	}

	while (pe != e && (bytes = write(f, pe, e - pe)) != 0) {
		if (bytes == -1) {
			if (errno == EINTR)
				continue;
			else
				die_x(EXIT_FAILURE, errno, config->result_file,
				      "Input file write error");
		}
		pe += bytes;
	}

	close(f);

	die_x(EXIT_SUCCESS, 0, config->result_file, result_ok);
}
